{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f37023d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ee603f2",
   "metadata": {},
   "source": [
    "When creating a custom exception in Python, it is recommended to inherit from the Exception class or one of its subclasses. The Exception class serves as the base class for all built-in exceptions in Python. Here's why we use the Exception class as the base class for custom exceptions:\n",
    "\n",
    "1. `Consistency and Compatibility`: By inheriting from the Exception class, your custom exception becomes compatible with the existing exception handling mechanisms in Python. It ensures that your custom exception can be caught and handled using the same exception handling syntax (try-except) that is used for built-in exceptions.<br>\n",
    "\n",
    "\n",
    "2. `Exception Hierarchy`: The Exception class is part of a hierarchical structure of built-in exception classes in Python. It provides a unified base for all exceptions, allowing you to create custom exceptions that fit into the existing hierarchy. This helps in organizing and categorizing exceptions based on their nature and relationship to other exceptions.<br>\n",
    "\n",
    "\n",
    "3. `Exception Handling`: Inheriting from the Exception class allows you to leverage the existing exception handling features in Python. You can catch your custom exception using a specific except block or a more general except block for handling a group of related exceptions. You can also catch multiple exceptions, including your custom exception, in a single except block.<br>\n",
    "\n",
    "\n",
    "4. `Customization and Specialization`: Inheriting from the Exception class allows you to customize and specialize your custom exception based on your specific requirements. You can define additional attributes and methods in your custom exception class to provide additional information or behavior relevant to your exception."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd184672",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Q2. Write a python program to print Python Exception Hierarchy.\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4c14349",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception\n",
      "    TypeError\n",
      "        FloatOperation\n",
      "        MultipartConversionError\n",
      "    StopAsyncIteration\n",
      "    StopIteration\n",
      "    ImportError\n",
      "        ModuleNotFoundError\n",
      "            PackageNotFoundError\n",
      "            PackageNotFoundError\n",
      "        ZipImportError\n",
      "    OSError\n",
      "        ConnectionError\n",
      "            BrokenPipeError\n",
      "            ConnectionAbortedError\n",
      "            ConnectionRefusedError\n",
      "            ConnectionResetError\n",
      "                RemoteDisconnected\n",
      "        BlockingIOError\n",
      "        ChildProcessError\n",
      "        FileExistsError\n",
      "        FileNotFoundError\n",
      "        IsADirectoryError\n",
      "        NotADirectoryError\n",
      "        InterruptedError\n",
      "            InterruptedSystemCall\n",
      "        PermissionError\n",
      "        ProcessLookupError\n",
      "        TimeoutError\n",
      "        UnsupportedOperation\n",
      "        ItimerError\n",
      "        herror\n",
      "        gaierror\n",
      "        timeout\n",
      "        SSLError\n",
      "            SSLCertVerificationError\n",
      "            SSLZeroReturnError\n",
      "            SSLWantReadError\n",
      "            SSLWantWriteError\n",
      "            SSLSyscallError\n",
      "            SSLEOFError\n",
      "        Error\n",
      "            SameFileError\n",
      "        SpecialFileError\n",
      "        ExecError\n",
      "        ReadError\n",
      "        URLError\n",
      "            HTTPError\n",
      "            ContentTooShortError\n",
      "        BadGzipFile\n",
      "    EOFError\n",
      "        IncompleteReadError\n",
      "    RuntimeError\n",
      "        RecursionError\n",
      "        NotImplementedError\n",
      "            ZMQVersionError\n",
      "            StdinNotImplementedError\n",
      "        _DeadlockError\n",
      "        BrokenBarrierError\n",
      "        BrokenExecutor\n",
      "            BrokenThreadPool\n",
      "        SendfileNotAvailableError\n",
      "        ExtractionError\n",
      "        VariableError\n",
      "        RemovedCommandError\n",
      "        PackageDiscoveryError\n",
      "    NameError\n",
      "        UnboundLocalError\n",
      "    AttributeError\n",
      "        FrozenInstanceError\n",
      "    SyntaxError\n",
      "        IndentationError\n",
      "            TabError\n",
      "    LookupError\n",
      "        IndexError\n",
      "        KeyError\n",
      "            NoSuchKernel\n",
      "            UnknownBackend\n",
      "        CodecRegistryError\n",
      "    ValueError\n",
      "        UnicodeError\n",
      "            UnicodeEncodeError\n",
      "            UnicodeDecodeError\n",
      "            UnicodeTranslateError\n",
      "        UnsupportedOperation\n",
      "        JSONDecodeError\n",
      "        SSLCertVerificationError\n",
      "        Error\n",
      "        IllegalMonthError\n",
      "        IllegalWeekdayError\n",
      "        ParserError\n",
      "        ClassNotFound\n",
      "        ClipboardEmpty\n",
      "        MessageDefect\n",
      "            NoBoundaryInMultipartDefect\n",
      "            StartBoundaryNotFoundDefect\n",
      "            CloseBoundaryNotFoundDefect\n",
      "            FirstHeaderLineIsContinuationDefect\n",
      "            MisplacedEnvelopeHeaderDefect\n",
      "            MissingHeaderBodySeparatorDefect\n",
      "            MultipartInvariantViolationDefect\n",
      "            InvalidMultipartContentTransferEncodingDefect\n",
      "            UndecodableBytesDefect\n",
      "            InvalidBase64PaddingDefect\n",
      "            InvalidBase64CharactersDefect\n",
      "            InvalidBase64LengthDefect\n",
      "            HeaderDefect\n",
      "                InvalidHeaderDefect\n",
      "                HeaderMissingRequiredValue\n",
      "                NonPrintableDefect\n",
      "                ObsoleteHeaderDefect\n",
      "                NonASCIILocalPartDefect\n",
      "        MacroToEdit\n",
      "        InvalidFileException\n",
      "        UnequalIterablesError\n",
      "        InvalidVersion\n",
      "        _InvalidELFFileHeader\n",
      "        InvalidWheelFilename\n",
      "        InvalidSdistFilename\n",
      "        InvalidSpecifier\n",
      "        InvalidMarker\n",
      "        UndefinedComparison\n",
      "        UndefinedEnvironmentName\n",
      "        InvalidRequirement\n",
      "            RequirementParseError\n",
      "        InvalidVersion\n",
      "        UnequalIterablesError\n",
      "        InvalidVersion\n",
      "        _InvalidELFFileHeader\n",
      "        InvalidWheelFilename\n",
      "        InvalidSdistFilename\n",
      "        InvalidSpecifier\n",
      "    AssertionError\n",
      "    ArithmeticError\n",
      "        FloatingPointError\n",
      "        OverflowError\n",
      "        ZeroDivisionError\n",
      "            DivisionByZero\n",
      "            DivisionUndefined\n",
      "        DecimalException\n",
      "            Clamped\n",
      "            Rounded\n",
      "                Underflow\n",
      "                Overflow\n",
      "            Inexact\n",
      "                Underflow\n",
      "                Overflow\n",
      "            Subnormal\n",
      "                Underflow\n",
      "            DivisionByZero\n",
      "            FloatOperation\n",
      "            InvalidOperation\n",
      "                ConversionSyntax\n",
      "                DivisionImpossible\n",
      "                DivisionUndefined\n",
      "                InvalidContext\n",
      "    SystemError\n",
      "        CodecRegistryError\n",
      "    ReferenceError\n",
      "    MemoryError\n",
      "    BufferError\n",
      "    Warning\n",
      "        UserWarning\n",
      "            GetPassWarning\n",
      "            FormatterWarning\n",
      "            _WouldIgnoreField\n",
      "            _BetaConfiguration\n",
      "            _InvalidFile\n",
      "        DeprecationWarning\n",
      "            ProvisionalWarning\n",
      "        PendingDeprecationWarning\n",
      "        SyntaxWarning\n",
      "        RuntimeWarning\n",
      "            ProactorSelectorThreadWarning\n",
      "            UnknownTimezoneWarning\n",
      "            PEP440Warning\n",
      "        FutureWarning\n",
      "            ProvisionalCompleterWarning\n",
      "        ImportWarning\n",
      "        UnicodeWarning\n",
      "        BytesWarning\n",
      "        ResourceWarning\n",
      "        DeprecatedTzFormatWarning\n",
      "        PkgResourcesDeprecationWarning\n",
      "        SetuptoolsDeprecationWarning\n",
      "            DistDeprecationWarning\n",
      "    Error\n",
      "    _OptionError\n",
      "    error\n",
      "    Verbose\n",
      "    _Error\n",
      "    SubprocessError\n",
      "        CalledProcessError\n",
      "        TimeoutExpired\n",
      "    TokenError\n",
      "    StopTokenizing\n",
      "    ClassFoundException\n",
      "    EndOfBlock\n",
      "    TraitError\n",
      "    Error\n",
      "    Error\n",
      "        CancelledError\n",
      "        TimeoutError\n",
      "        InvalidStateError\n",
      "    _GiveupOnSendfile\n",
      "    error\n",
      "    Incomplete\n",
      "    TimeoutError\n",
      "    InvalidStateError\n",
      "    LimitOverrunError\n",
      "    QueueEmpty\n",
      "    QueueFull\n",
      "    Empty\n",
      "    Full\n",
      "    ArgumentError\n",
      "    ZMQBaseError\n",
      "        ZMQError\n",
      "            ContextTerminated\n",
      "            Again\n",
      "            InterruptedSystemCall\n",
      "        ZMQBindError\n",
      "        NotDone\n",
      "    PickleError\n",
      "        PicklingError\n",
      "        UnpicklingError\n",
      "    _Stop\n",
      "    ArgumentError\n",
      "    ArgumentTypeError\n",
      "    ConfigError\n",
      "        ConfigLoaderError\n",
      "            ArgumentError\n",
      "        ConfigFileNotFound\n",
      "    ConfigurableError\n",
      "        MultipleInstanceError\n",
      "    ApplicationError\n",
      "    error\n",
      "    TimeoutError\n",
      "    error\n",
      "    ReturnValueIgnoredError\n",
      "    KeyReuseError\n",
      "    UnknownKeyError\n",
      "    LeakedCallbackError\n",
      "    BadYieldError\n",
      "    ReturnValueIgnoredError\n",
      "    Return\n",
      "    InvalidPortNumber\n",
      "    error\n",
      "    LZMAError\n",
      "    RegistryError\n",
      "    _GiveupOnFastCopy\n",
      "    NoIPAddresses\n",
      "    BadZipFile\n",
      "    LargeZipFile\n",
      "    Error\n",
      "        NoSectionError\n",
      "        DuplicateSectionError\n",
      "        DuplicateOptionError\n",
      "        NoOptionError\n",
      "        InterpolationError\n",
      "            InterpolationMissingOptionError\n",
      "            InterpolationSyntaxError\n",
      "            InterpolationDepthError\n",
      "        ParsingError\n",
      "            MissingSectionHeaderError\n",
      "    BadEntryPoint\n",
      "    NoSuchEntryPoint\n",
      "    DuplicateKernelError\n",
      "    ErrorDuringImport\n",
      "    BdbQuit\n",
      "    Restart\n",
      "    ExceptionPexpect\n",
      "        EOF\n",
      "        TIMEOUT\n",
      "    PtyProcessError\n",
      "    FindCmdError\n",
      "    HomeDirError\n",
      "    ProfileDirError\n",
      "    IPythonCoreError\n",
      "        TryNext\n",
      "        UsageError\n",
      "        StdinNotImplementedError\n",
      "    InputRejected\n",
      "    GetoptError\n",
      "    OptionError\n",
      "    ErrorToken\n",
      "    PrefilterError\n",
      "    AliasError\n",
      "        InvalidAliasError\n",
      "    Error\n",
      "        InterfaceError\n",
      "        DatabaseError\n",
      "            InternalError\n",
      "            OperationalError\n",
      "            ProgrammingError\n",
      "            IntegrityError\n",
      "            DataError\n",
      "            NotSupportedError\n",
      "    Warning\n",
      "    SpaceInInput\n",
      "    DOMException\n",
      "        IndexSizeErr\n",
      "        DomstringSizeErr\n",
      "        HierarchyRequestErr\n",
      "        WrongDocumentErr\n",
      "        InvalidCharacterErr\n",
      "        NoDataAllowedErr\n",
      "        NoModificationAllowedErr\n",
      "        NotFoundErr\n",
      "        NotSupportedErr\n",
      "        InuseAttributeErr\n",
      "        InvalidStateErr\n",
      "        SyntaxErr\n",
      "        InvalidModificationErr\n",
      "        NamespaceErr\n",
      "        InvalidAccessErr\n",
      "        ValidationErr\n",
      "    ValidationError\n",
      "    EditReadOnlyBuffer\n",
      "    _Retry\n",
      "    InvalidLayoutError\n",
      "    HeightIsUnknownError\n",
      "    ParserSyntaxError\n",
      "    InternalParseError\n",
      "    _PositionUpdatingFinished\n",
      "    SimpleGetItemNotFound\n",
      "    UncaughtAttributeError\n",
      "    HasNoContext\n",
      "    ParamIssue\n",
      "    _JediError\n",
      "        InternalError\n",
      "        WrongVersion\n",
      "        RefactoringError\n",
      "    OnErrorLeaf\n",
      "    InvalidPythonEnvironment\n",
      "    MessageError\n",
      "        MessageParseError\n",
      "            HeaderParseError\n",
      "                _InvalidEwError\n",
      "            BoundaryError\n",
      "        MultipartConversionError\n",
      "        CharsetError\n",
      "    Error\n",
      "    HTTPException\n",
      "        NotConnected\n",
      "        InvalidURL\n",
      "        UnknownProtocol\n",
      "        UnknownTransferEncoding\n",
      "        UnimplementedFileMode\n",
      "        IncompleteRead\n",
      "        ImproperConnectionState\n",
      "            CannotSendRequest\n",
      "            CannotSendHeader\n",
      "            ResponseNotReady\n",
      "        BadStatusLine\n",
      "            RemoteDisconnected\n",
      "        LineTooLong\n",
      "    InteractivelyDefined\n",
      "    KillEmbedded\n",
      "    Error\n",
      "        NoSuchProcess\n",
      "            ZombieProcess\n",
      "        AccessDenied\n",
      "        TimeoutExpired\n",
      "    ZombieProcessError\n",
      "    QueueEmpty\n",
      "    QueueFull\n",
      "    ExpatError\n",
      "    Error\n",
      "        ProtocolError\n",
      "        ResponseError\n",
      "        Fault\n",
      "    ParseBaseException\n",
      "        ParseException\n",
      "        ParseFatalException\n",
      "            ParseSyntaxException\n",
      "    RecursiveGrammarException\n",
      "    ResolutionError\n",
      "        VersionConflict\n",
      "            ContextualVersionConflict\n",
      "        DistributionNotFound\n",
      "        UnknownExtra\n",
      "    UnableToResolveVariableException\n",
      "    InvalidTypeInArgsException\n",
      "    DistutilsError\n",
      "        DistutilsModuleError\n",
      "        DistutilsClassError\n",
      "        DistutilsGetoptError\n",
      "        DistutilsArgError\n",
      "        DistutilsFileError\n",
      "        DistutilsOptionError\n",
      "        DistutilsSetupError\n",
      "        DistutilsPlatformError\n",
      "        DistutilsExecError\n",
      "        DistutilsInternalError\n",
      "        DistutilsTemplateError\n",
      "        DistutilsByteCompileError\n",
      "        RemovedCommandError\n",
      "        PackageDiscoveryError\n",
      "    CCompilerError\n",
      "        PreprocessError\n",
      "        CompileError\n",
      "        LibError\n",
      "        LinkError\n",
      "        UnknownFileError\n",
      "    Error\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy(exception_class, indent=0):\n",
    "    print(' ' * indent + exception_class.__name__)\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 4)\n",
    "\n",
    "# Print the exception hierarchy starting from the Exception class\n",
    "print_exception_hierarchy(Exception)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7008ee73",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Q3. What errors are defined in the ArithmeticError class? Explain any two with an example.\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c682358e",
   "metadata": {},
   "source": [
    "The ArithmeticError class in Python is the base class for exceptions related to arithmetic operations. It serves as a superclass for more specific arithmetic-related exceptions. Here are two commonly used exceptions derived from the ArithmeticError class along with examples:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "916655cd",
   "metadata": {},
   "source": [
    "`ZeroDivisionError`: This exception is raised when there is an attempt to divide a number by zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef0d72d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: division by zero\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    result = 10 / 0  # Attempting to divide by zero\n",
    "except ZeroDivisionError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ee10c0d",
   "metadata": {},
   "source": [
    "`OverflowError`: This exception is raised when the result of an arithmetic operation exceeds the maximum representable value.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ca75c90d",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    result = 2 ** 1000  # Calculating 2 raised to the power of 1000\n",
    "except OverflowError as e:\n",
    "    print(\"Error:\", e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96995f67",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Q4. Why LookupError class is used? Explain with an example KeyError and IndexError.\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9e22b52",
   "metadata": {},
   "source": [
    "The `LookupError` class in Python is the base class for exceptions that occur when a lookup or indexing operation fails. It serves as a superclass for more specific lookup-related exceptions. Here are two commonly used exceptions derived from the LookupError class along with examples:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc79b71c",
   "metadata": {},
   "source": [
    "1. `KeyError`: This exception is raised when a dictionary key or a set element is not found.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0bf0f1f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: 'd'\n"
     ]
    }
   ],
   "source": [
    "my_dict = {'a': 1, 'b': 2, 'c': 3}\n",
    "\n",
    "try:\n",
    "    value = my_dict['d']  # Accessing a non-existent key 'd'\n",
    "except KeyError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de178f2b",
   "metadata": {},
   "source": [
    "2. `IndexError`: This exception is raised when an index is out of range in a sequence (such as a list or a string)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "16847c36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: list index out of range\n"
     ]
    }
   ],
   "source": [
    "my_list = [1, 2, 3]\n",
    "\n",
    "try:\n",
    "    value = my_list[3]  # Accessing an index that is out of range\n",
    "except IndexError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13b8d87e",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Q5. Explain ImportError. What is ModuleNotFoundError?\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "095e09b9",
   "metadata": {},
   "source": [
    "ImportError is an exception class in Python that is raised when an import statement fails to find and import a module. It is a generic exception that encompasses various import-related errors. This exception can be raised in the following scenarios:\n",
    "\n",
    "1. The specified module does not exist.\n",
    "2. There is an issue with the module's code or structure.\n",
    "3. The module's dependencies are not installed or accessible.\n",
    "4. The module's name conflicts with another object in the namespace.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f9078c50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: No module named 'non_existent_module'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import non_existent_module  # Trying to import a non-existent module\n",
    "except ImportError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eac01051",
   "metadata": {},
   "source": [
    "On the other hand, `ModuleNotFoundError` is a subclass of ImportError that specifically indicates that the module being imported could not be found. It was introduced in Python 3.6 to provide a more specific error message for import failures due to module nonexistence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ffe936cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: No module named 'non_existent_module'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import non_existent_module  # Trying to import a non-existent module\n",
    "except ModuleNotFoundError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b807e075",
   "metadata": {},
   "source": [
    "As you can see, ModuleNotFoundError provides the same error message as ImportError in this case. However, using ModuleNotFoundError helps to make the error message more descriptive and distinguishes it from other possible import-related errors.\n",
    "\n",
    "Both ImportError and ModuleNotFoundError are raised when there are issues with importing a module, such as when the module is not found or cannot be loaded. They allow you to handle import failures and provide appropriate error handling or fallback strategies in your code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ffc9a41",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Q6. List down some best practices for exception handling in python.\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e61ffbfb",
   "metadata": {},
   "source": [
    "1. `Use specific exception types`: Catch specific exception types whenever possible instead of using a generic except block. This allows you to handle different exceptions differently and provide appropriate error messages or actions based on the specific type of exception.\n",
    "\n",
    "2. `Keep try blocks minimal`: Place only the necessary code within the try block that might raise an exception. Avoid placing large blocks of code within a single try block, as it can make it harder to pinpoint the exact location of the exception.\n",
    "\n",
    "3. `Use finally for cleanup`: Utilize the finally block to include cleanup code that should always execute, regardless of whether an exception is raised or not. This can be useful for releasing resources, closing files, or cleaning up connections.\n",
    "\n",
    "4. `Handle exceptions at the appropriate level`: Handle exceptions at a level where you have enough context to handle the error or take appropriate action. Avoid catching exceptions too early if you cannot handle them effectively at that point.\n",
    "\n",
    "5. `Avoid bare except statements`: Avoid using a bare except statement without specifying the exception type. It can make it harder to identify and debug issues since it catches all exceptions, including ones you might not have anticipated.\n",
    "\n",
    "6. `Use multiple except blocks`: Use multiple except blocks to handle different exceptions separately. This allows you to provide tailored error handling and appropriate actions based on the specific exception types.\n",
    "\n",
    "7. `Use else block when appropriate`: Use the else block after the try-except block to specify code that should execute only if no exceptions are raised. This can be useful for separating error handling code from normal execution flow.\n",
    "\n",
    "8. `Log or report exceptions`: Consider logging or reporting exceptions to aid in debugging and troubleshooting. You can use Python's built-in logging module or external logging frameworks to record exception details for later analysis.\n",
    "\n",
    "9. `Avoid unnecessary exceptions`: Avoid raising or catching exceptions unnecessarily. Use conditional statements and proper validation to prevent exceptions where possible.\n",
    "\n",
    "10. `Document exceptions`: Document the exceptions that can be raised by your code, especially for custom exceptions and public interfaces. This helps users of your code to handle exceptions appropriately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "185045c9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
